### ISSUE TITLE
Consider adding a function to do sum/max/min/avg/count/stddev at once

### ISSUE THREAD

**Issue by alburthoffman** [prometheus/prometheus#8199]
It's quite normal that doing spatial aggregations in prometheus rules. For different metrics different group labels are needed. So basically if want to get sum/max/min/avg/count/stddev for all these metrics, we need 6 rules at present. Consider if we have more than 1000 metrics to aggregate, then there would be about 6000 rules, which are a lot.

But this could be simply using one function, like sparial_aggregate(vector, "key1,key2", "category") or sparial_aggregate(vector, "key1,key2", "category", "sum, count, avg"):
the function will caculate the results like sum/max/min/avg/count/stddev and add category as an label to indicate which type of aggregations one series has.

The benefit is we only needs 1000 rules, and each rule only query one dataset and outputs all.

It's hard to do this via operator, maybe it can be done like below:
`aggall by (key1,key2) ()`
and aggall will use one preserve label to identify the results, like `_category_`

**Comment by brian-brazil** [comment]
> add category as an label to indicate which type of aggregations one series has.

This would break the rule that a metric should make sense when averaged or summed. This could result in situations where cardinality is increased 6x compared to the input expression which is asking for problems, as it would allow cardinality explosions within PromQL which are not possible today. The number of rules is not a concern, if you have 1k identical metrics and rules (which is a lot already) then you almost certainly have templating generating them.

I don't think this makes sense.

**Comment by alburthoffman** [comment]
The 6x cardinality is expected, and it still exists when defining 6 recording rules independently. 

The template helps a little bit. but the rules are mainly created by other teams, not our self. so it is not easy to maintain the templates, like you have to deleting or updating. it actually raise other problems.

All our users want something like aggregations by given labels, and if one function works, then it reduces their efforts. Also a happy for our system as well. 

because most likely 1 or 2 prometheus can handle 1k rules. But for 6k rules, there would be 6x capacities.

**Comment by alburthoffman** [comment]
it might be wired to add something like category label when aggregating. But it might be other better solution for this, someone else can provide if they have.

**Comment by RalphSu** [comment]
If user decide to use agg function, then it's is aware that multiple aggregation are asked. Etc, in site monitoring, people looking at those aggregation of golden signals (https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#:~:text=The%20four%20golden%20signals%20of,traffic%2C%20errors%2C%20and%20saturation.&text=Therefore%2C%20it's%20important%20to%20track,%2Dlevel%20system%2Dspecific%20metric.)

Another point is that, generate the rules are not a major concern, since they are easier enough to build templating. But 6x rules means 6X query to data store, at 200K scale of golden signals, 5 minutes windows, aggregation in one query seems to be worth?

**Comment by roidelapluie** [comment]
We have looked at this during our bug scrub, and it seems that this can be solved by aligning metric names accross jobs, and then just have 6 rules for all your jobs. This is not in line with how you would use metrics and the prometheus data model.

Therefore we have decided to close this issue. Thanks!

**Comment by bwplotka** [comment]
Some 2p from me: This is generally a bad idea. (:

We did that once at my previous job: We had a template language for our Prometheus Recording Rules (`m4` - really bad one ;p). What we did is that we had a function like this and it was generating ~6 recording rules per each invocation to satisfy all aggregations.

It worked, but it caused huge cardinality spam (6x) without **a good reason**.

It was very convenient to create such rules but in the end, only `sum` was used and only on some metrics, leaving millions of series never used and expensive to look up. Cleaning those up was incredibly hard because it's not trivial to say what recording rules were used, which were not.

What I would recommend instead is... Thanos and its downsampling. Downsampling is exactly this. All aggregations over time, without adding extra index on top.
